scapy/contrib/pnio_dcp.py

```diff
@@ -0,0 +1,607 @@
+#####################################################
+
+DCP_GET_SET_FRAME_ID = 0xFEFD
+DCP_IDENTIFY_REQUEST_FRAME_ID = 0xFEFE
+DCP_IDENTIFY_RESPONSE_FRAME_ID = 0xFEFF
+
+DCP_REQUEST = 0x00
+DCP_RESPONSE = 0x01
+
+DCP_SERVICE_ID_GET = 0x03
+DCP_SERVICE_ID_SET = 0x04
+DCP_SERVICE_ID_IDENTIFY = 0x05
+
+DCP_SERVICE_ID = {
+    0x00: "reserved",
+    0x01: "Manufacturer specific",
+    0x02: "Manufacturer specific",
+    0x03: "Get",
+    0x04: "Set",
+    0x05: "Identify",
+    0x06: "Hello",
+}
+
+DCP_SERVICE_TYPE = {
+    0x00: "Request",
+    0x01: "Response Success",
+    0x05: "Response - Request not supported",
+}
+
+DCP_DEVICE_ROLES = {
+    0x00: "IO Supervisor",
+    0x01: "IO Device",
+    0x02: "IO Controller",
+
+}
+
+DCP_OPTIONS = {
+    0x00: "reserved",
+    0x01: "IP",
+    0x02: "Device properties",
+    0x03: "DHCP",
+    0x04: "Reserved",
+    0x05: "Control",
+    0x06: "Device Initiative",
+    0xff: "All Selector"
+}
+reserved = {i: "reserved" for i in range(0x07, 0x7f)}
+manufacturer_specific = {i: "Manufacturer specific" for i in range(0x80, 0xfe)}
+DCP_OPTIONS.update(reserved)
+DCP_OPTIONS.update(manufacturer_specific)
+
+DCP_SUBOPTIONS = {
+    # ip
+    0x01: {
+        0x00: "Reserved",
+        0x01: "MAC Address",
+        0x02: "IP Parameter"
+    },
+    # device properties
+    0x02: {
+        0x00: "Reserved",
+        0x01: "Manufacturer specific (Type of Station)",
+        0x02: "Name of Station",
+        0x03: "Device ID",
+        0x04: "Device Role",
+        0x05: "Device Options",
+        0x06: "Alias Name",
+        0x07: "Device Instance",
+        0x08: "OEM Device ID",
+    },
+    # dhcp
+    0x03: {
+        0x0c: "Host name",
+        0x2b: "Vendor specific",
+        0x36: "Server identifier",
+        0x37: "Parameter request list",
+        0x3c: "Class identifier",
+        0x3d: "DHCP client identifier",
+        0x51: "FQDN, Fully Qualified Domain Name",
+        0x61: "UUID/GUID-based Client",
+        0xff: "Control DHCP for address resolution"
+    },
+    # control
+    0x05: {
+        0x00: "Reserved",
+        0x01: "Start Transaction",
+        0x02: "End Transaction",
+        0x03: "Signal",
+        0x04: "Response",
+        0x05: "Reset Factory Settings",
+        0x06: "Reset to Factory"
+    },
+    # device initiative
+    0x06: {
+        0x00: "Reserved",
+        0x01: "Device Initiative"
+    },
+    0xff: {
+        0xff: "ALL Selector"
+    }
+}
+
+BLOCK_INFOS = {
+    0x00: "Reserved",
+}
+reserved = {i: "reserved" for i in range(0x01, 0xff)}
+BLOCK_INFOS.update(reserved)
+
+
+IP_BLOCK_INFOS = {
+    0x0000: "IP not set",
+    0x0001: "IP set",
+    0x0002: "IP set by DHCP",
+    0x0080: "IP not set (address conflict detected)",
+    0x0081: "IP set (address conflict detected)",
+    0x0082: "IP set by DHCP (address conflict detected)",
+}
+reserved = {i: "reserved" for i in range(0x0003, 0x007f)}
+IP_BLOCK_INFOS.update(reserved)
```

Comment:
Could you do that in one shot (without even the dictionary), to use a generator and not store a useless dictionary ?

The same applies elsewhere on the file

Reply:
Hi Gabriel,
honestly, I don't know how to do that.
I was looking for examples from other authors and found one in 'scapy/contrib/scada/iec104'. 

See l. 118 _def _dict_add_reserved_range(d, start, end):_

Should I implement it in that way?

Best regards

Reply:
Sorry, I just meant
```
IP_BLOCK_INFOS.update({i: "reserved" for i in range(0x0003, 0x007f)})
```
So that you don't keep a `reserved` dictionnary